File: codeserver/img2img.java
/*
 * img2img.java
 *
 * Compile:
 *   javac img2img.java
 *
 * Usage:
 *   java img2img <directory> [--type png|jpeg|jpg|...] [--del]
 *
 * Examples:
 *   java img2img myimages
 *   java img2img myimages --type jpeg
 *   java img2img myimages --type jpg --del
 *
 * - Converts all compatible images in <directory> (recursively) to the specified type (default: png).
 * - Outputs converted images to the 'out' directory.
 * - Logs each run and per-image results in out/skipped.txt.
 * - Use --del to delete source images after conversion.
 */

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import javax.imageio.ImageIO;

public class img2img {
    private static final List<String> IMAGE_EXTENSIONS = Arrays.asList(
        ".jpg", ".jpeg", ".bmp", ".gif", ".tiff", ".webp", ".jfif"
    );
    private static final String OUT_DIR = "out";
    private static final String LOG_FILE = "skipped.txt";

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java img2img <directory> [--type png|jpeg|jpg|...] [--del]");
            return;
        }
        String directory = args[0];
        String outType = "png";
        boolean deleteSource = false;
        for (int i = 1; i < args.length; i++) {
            if (args[i].equals("--type") && i + 1 < args.length) {
                outType = args[i + 1].toLowerCase();
                i++;
            } else if (args[i].equals("--del")) {
                deleteSource = true;
            }
        }
        convertImages(directory, OUT_DIR, outType, deleteSource);
    }

    private static void convertImages(String directory, String outputDir, String outType, boolean deleteSource) {
        List<String> compatible = new ArrayList<>();
        List<String> nonCompatible = new ArrayList<>();
        getCompatibleImages(directory, compatible, nonCompatible);
        int total = compatible.size();
        File outDir = new File(outputDir);
        if (!outDir.exists()) outDir.mkdirs();
        File logFile = new File(outDir, LOG_FILE);
        String now = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
        try {
            Files.write(logFile.toPath(), ("\n--- Run at " + now + " ---\n").getBytes(), Files.exists(logFile.toPath()) ? java.nio.file.StandardOpenOption.APPEND : java.nio.file.StandardOpenOption.CREATE);
            Files.write(logFile.toPath(), ("Directory: " + directory + "\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
            Files.write(logFile.toPath(), ("Output type: " + outType + "\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
            Files.write(logFile.toPath(), ("Delete source: " + deleteSource + "\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
            Files.write(logFile.toPath(), ("Found " + total + " compatible images.\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
            if (!nonCompatible.isEmpty()) {
                Files.write(logFile.toPath(), "Non-compatible files:\n".getBytes(), java.nio.file.StandardOpenOption.APPEND);
                for (String item : nonCompatible) {
                    Files.write(logFile.toPath(), ("  " + item + "\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
                }
            } else {
                Files.write(logFile.toPath(), "No non-compatible files.\n".getBytes(), java.nio.file.StandardOpenOption.APPEND);
            }
        } catch (IOException e) {
            System.out.println("Failed to write log: " + e.getMessage());
        }
        if (total == 0) {
            System.out.println("No compatible images found.");
            System.out.println("Non-compatible files: " + nonCompatible);
            System.out.println("Run log written to: " + logFile.getAbsolutePath());
            return;
        }
        System.out.println("Found " + total + " compatible images.");
        if (!nonCompatible.isEmpty()) {
            System.out.println("Non-compatible files:");
            for (String item : nonCompatible) {
                System.out.println("  " + item);
            }
            System.out.println("Run log written to: " + logFile.getAbsolutePath());
        }
        int idx = 1;
        for (String imgPath : compatible) {
            String baseName = new File(imgPath).getName();
            int dot = baseName.lastIndexOf('.');
            if (dot > 0) baseName = baseName.substring(0, dot);
            String outExt = outType.equals("jpg") ? ".jpeg" : "." + outType;
            String outFile = outputDir + File.separator + baseName + outExt;
            String saveType = outType.equals("jpg") ? "jpeg" : outType;
            try {
                BufferedImage img = ImageIO.read(new File(imgPath));
                if (img == null) throw new IOException("Not a supported image file");
                ImageIO.write(img, saveType, new File(outFile));
                appendLog(logFile, "SUCCESS: " + imgPath + " -> " + outFile);
                if (deleteSource) {
                    try {
                        Files.delete(Paths.get(imgPath));
                        appendLog(logFile, "DELETED: " + imgPath);
                    } catch (IOException delE) {
                        appendLog(logFile, "FAILED TO DELETE: " + imgPath + " | " + delE.getMessage());
                    }
                }
            } catch (Exception e) {
                appendLog(logFile, "FAILED: " + imgPath + " | " + e.getMessage());
                System.out.println("Failed to convert " + imgPath + ": " + e.getMessage());
            }
            printProgress(idx++, total, imgPath);
        }
        System.out.println("Conversion complete.");
    }

    private static void getCompatibleImages(String directory, List<String> compatible, List<String> nonCompatible) {
        File dir = new File(directory);
        if (!dir.exists() || !dir.isDirectory()) return;
        for (File file : dir.listFiles()) {
            if (file.isDirectory()) {
                getCompatibleImages(file.getAbsolutePath(), compatible, nonCompatible);
            } else {
                String ext = getFileExtension(file.getName());
                if (IMAGE_EXTENSIONS.contains(ext)) {
                    compatible.add(file.getAbsolutePath());
                } else {
                    nonCompatible.add(file.getAbsolutePath());
                }
            }
        }
    }

    private static String getFileExtension(String filename) {
        int dot = filename.lastIndexOf('.');
        return dot >= 0 ? filename.substring(dot).toLowerCase() : "";
    }

    private static void appendLog(File logFile, String line) {
        try {
            Files.write(logFile.toPath(), (line + "\n").getBytes(), java.nio.file.StandardOpenOption.APPEND);
        } catch (IOException e) {
            System.out.println("Failed to write log: " + e.getMessage());
        }
    }

    private static void printProgress(int current, int total, String filename) {
        int barLength = 40;
        double percent = (double) current / total;
        int filled = (int) (percent * barLength);
        String bar = "[" + "-".repeat(Math.max(0, filled - 1)) + (filled > 0 ? ">" : "") + " ".repeat(barLength - filled) + "] ";
        System.out.print("\r" + bar + current + "/" + total + "  " + new File(filename).getName());
        if (current == total) System.out.println();
    }
}

File: codeserver/img2img.py
"""
img2img.py

Usage:
  python img2img.py <directory> [--type png|jpeg|jpg|...] [--del]

Examples:
  python img2img.py myimages
  python img2img.py myimages --type jpeg
  python img2img.py myimages --type jpg --del

- Converts all compatible images in <directory> (recursively) to the specified type (default: png).
- Outputs converted images to the 'out' directory.
- Logs each run and per-image results in out/skipped.txt.
- Use --del to delete source images after conversion.
- Requires: Pillow, tqdm
  pip install pillow tqdm
"""

import os
from PIL import Image
import sys
import shutil
import argparse
from tqdm import tqdm
import datetime

# Supported image extensions
IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.bmp', '.gif', '.tiff', '.webp', '.jfif']

_dir = os.path.dirname(os.path.abspath(__file__))
out = os.path.join(_dir, 'out')

def get_compatible_images(directory):
    compatible = []
    non_compatible = []
    for root, _, files in os.walk(directory):
        for file in files:
            ext = os.path.splitext(file)[1].lower()
            path = os.path.join(root, file)
            if ext in IMAGE_EXTENSIONS:
                compatible.append(path)
            else:
                non_compatible.append(path)
    return compatible, non_compatible

def convert_images_to_png(directory, output_dir=out, out_type='png', delete_source=False):
    os.makedirs(output_dir, exist_ok=True)
    images, non_compatible = get_compatible_images(directory)
    total = len(images)
    log_path = os.path.join(output_dir, 'skipped.txt')
    now = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    with open(log_path, 'a') as f:
        f.write(f"\n--- Run at {now} ---\n")
        f.write(f"Directory: {directory}\n")
        f.write(f"Output type: {out_type}\n")
        f.write(f"Delete source: {delete_source}\n")
        f.write(f"Found {total} compatible images.\n")
        if non_compatible:
            f.write("Non-compatible files:\n")
            for item in non_compatible:
                f.write(f"  {item}\n")
        else:
            f.write("No non-compatible files.\n")
    if total == 0:
        print("No compatible images found.")
        print(f"Non-compatible files: {non_compatible}")
        print(f"Run log written to: {log_path}")
        return
    print(f"Found {total} compatible images.")
    if non_compatible:
        print(f"Non-compatible files:")
        for item in non_compatible:
            print(f"  {item}")
        print(f"Run log written to: {log_path}")
    with open(log_path, 'a') as f:
        for idx, img_path in enumerate(tqdm(images, desc='Converting', unit='img'), 1):
            try:
                with Image.open(img_path) as img:
                    base_name = os.path.splitext(os.path.basename(img_path))[0]
                    out_ext = '.' + out_type.lower().replace('jpg', 'jpeg') if out_type.lower() == 'jpg' else '.' + out_type.lower()
                    save_type = out_type.upper() if out_type.lower() != 'jpg' else 'JPEG'
                    out_file = os.path.join(output_dir, base_name + out_ext)
                    img.save(out_file, save_type)
                    f.write(f"SUCCESS: {img_path} -> {out_file}\n")
                if delete_source:
                    try:
                        os.remove(img_path)
                        f.write(f"DELETED: {img_path}\n")
                    except Exception as del_e:
                        f.write(f"FAILED TO DELETE: {img_path} | {del_e}\n")
            except Exception as e:
                f.write(f"FAILED: {img_path} | {e}\n")
                print(f"\nFailed to convert {img_path}: {e}")
    print("Conversion complete.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Convert images in a directory to a specified format.')
    parser.add_argument('directory', help='Directory to search for images')
    parser.add_argument('--type', default='png', help='Output image type (png, jpeg, jpg, etc). Default: png')
    parser.add_argument('--del', dest='delete_source', action='store_true', help='Delete source images after conversion')
    args = parser.parse_args()
    convert_images_to_png(args.directory, out, args.type, args.delete_source)
